src/index.ml

```diff
@@ -506,16 +514,10 @@ struct
       @~ find_if_exists ~name:"index" ~find:interpolation_search t.index
     in
     Mutex.with_lock t.rename_lock (fun () ->
-        if t.config.readonly then sync_log t;
         find_if_exists ~name:"log_async"
           ~find:(fun log -> Tbl.find log.mem)
           t.log_async
-        @~ fun () ->
-        find_log_index @~ fun () ->
-        if t.config.readonly then (
-          sync_log t;
-          find_log_index () )
-        else raise Not_found)
+        @~ find_log_index)
```

Comment:
The case that was handled by this retry is now completely omitted.
That means that it is virtually possible to call `ro_sync` and have the same interleaving behavior with `merge` as before, causing the `sync_log` to not see the disk modifications (that's why we retried before).

Reply:
Yes, and it's tricky, because no test fails (I run them 500 times). 
 
I looked at what interleaving execution was causing this bug, 
(https://github.com/mirage/index/pull/118#issuecomment-548083904), and this particular bug can be fixed at the end of a merge, by first setting the generation number and then clearing the log. It can cause a ro_sync to refill a log, even if there are no new values, but it cannot lead to a value not found anymore. What do you think?

Reply:
As discussed offline: we can test this properly with a couple of extra hooks, and we probably should.

Reply:
I added the hooks and tested them in commit https://github.com/mirage/index/pull/175/commits/8d0f15bc1ac8ad1ebf8313f00b85e811e7e22168. 
There is one test that fails in that commit: the RO reads the generation number and the offset to decide if a change occurred. If between the two reads both the generation changed and the log was cleared then RO cannot detect a change (see test https://github.com/mirage/index/pull/175/commits/8d0f15bc1ac8ad1ebf8313f00b85e811e7e22168#diff-81aec069a46f58216cc2e7046f8d183dR349). 
The solution I propose in https://github.com/mirage/index/pull/175/commits/0af8704f7d3f3626f1d99ff883d3373d9fd718c6 is to group generation and offset reads/writes into a single read/write.

Reply:
I knew you had this somewhere, but didn't know where :) I added some of your modifications in the commit, and added you as co-author. 
I did not add the `fan_size` in the headers, as it is not necessary for the correctness of this PR.